Support -trait and -notrait for RunTest.exe

oocx opened the pull request:
RunTests has additional arguments to run tests
with specific traits. These arguments can also be
specified as build properties for BuildAndTest.proj.
Example: msbuild BuildAndTest.proj /p:Trait="Feature=Classification". 
I also refactored parsing of the command line arguments into it's own class.

Fix #6241

diryboy commented on src/Tools/Source/RunTests/Options.cs:
```diff
@@ -0,0 +1,86 @@
+
+        private static void PrintUsage()
+        {
+            Console.WriteLine("runtests [xunit-console-runner] [-test64] [-xml] [-trait:listOfTraitsToInclude] [-notrait:listOfTraitsToExclude] [assembly1] [assembly2] [...]");
```
Would be nice if example/separator included. Like `[-trait:name1=value1;..]`.

diryboy commented on src/Tools/Source/RunTests/Options.cs:
```diff
@@ -0,0 +1,86 @@
+                }
+                else if (current.StartsWith("-trait:", StringComparison.OrdinalIgnoreCase))
+                {
+                    opt.Trait = current.Substring(7);
```
If the argument is just `-trait:`, this crash, similar below.

oocx commented:
Thank you for your feedback.

I have added additional error handling for the command line arguments. Only handling the "trait:" case would only fix the problem in Options.cs, but still cause an exception later in the program as this would add "trait:" to the assemblies list and try to read it as a file.

I also changed the PrintUsage output so that it should be easier to see how multiple traits can be specified.

davkean commented:
Added the failure for unit64 to https://github.com/dotnet/roslyn/issues/6166.

retest prtest/win/dbg/unit64

davkean commented:
test eta please
retest this please

davkean commented:
tag @dotnet/roslyn-infrastructure  for review

oocx commented:
I've been thinking about this some more. With the additional validation I added in the last commit, the tests would not run at all if one of the specified test assemblies is missing. That's different than the old behavior. Should I leave it like that or remove the File.Exists? If I remove it, I would probably have to replace it with a check that verifies that the file name is at least valid, otherwise we would be back to the previous problem that invalid command line arguments like "-trait:" (without anything after the ':') are added to the assembly list.

davkean commented:
Adding @dotnet/roslyn-ide also to take a look.

dpoeschl commented:
Previous prtest/lin/dbg/unit32 test run failed (http://dotnet-ci.cloudapp.net/job/roslyn_prtest_lin_dbg_unit32/552/):

```
10:57:57 Native stacktrace:
10:57:57 
10:57:57 mono_class_from_mono_type: implement me 0x9d
10:57:57 
10:57:57 * Assertion: should not be reached at class.c:6523
10:57:57 
10:57:57 
./cibuild.sh: line 234: 17794 Aborted                 (core dumped) mono $MONO_ARGS $xunit_runner Binaries/$BUILD_CONFIGURATION/$i.dll -xml Binaries/$BUILD_CONFIGURATION/xUnitResults/$i.dll.xml -noshadow
```

Filed https://github.com/dotnet/roslyn/issues/6348 to track that. Let's see if it fails that way again...

retest prtest/lin/dbg/unit32 please

dpoeschl commented:
retest prtest/lin/dbg/unit32 please

dpoeschl commented:
retest prtest/lin/dbg/unit32 please

Pilchie commented on src/Tools/Source/RunTests/Options.cs:
```diff
@@ -0,0 +1,111 @@
+                    opt.UseHtml = false;
+                    index++;
+                }
+                else if (current.Length > 7 && current.StartsWith("-trait:", StringComparison.OrdinalIgnoreCase))
```
You could re-use `comp` here (or inline it everywhere).

oocx replied:
comp is a StringCompar**er**, StartsWith expects a StringCompar**ison**. I don't see how I could reuse comp with StartsWith.

Pilchie replied:
Ah yes.  Missed that, sorry.

oocx replied:
No problem :).

Pilchie commented:
Tagging @tannergooding @agocke @jaredpar explicitly.  What do you think about @oocx 's question about the `File.Exists` check?

ManishJayaswal commented:
@oocx I think it makes sense to run the rest of the tests even if one or more assemblies are missing. At the end we should report the missing assemblies.

oocx commented:
RunTests will now run even if some of the assemblies are missing / invalid. If assemblies are missing / invalid, it will output a list of the missing assemblies after running the tests.

davkean commented:
Thanks - sorry this got missed, can you resolve conflicts and resubmit - and we'll get this through.

oocx commented:
Done. I hope I merged correctly, I have never done that with git before.

davkean commented:
Merging this in, thanks for your contribution!

diryboy commented:
thanks!
